### PR TITLE
Use ld.gold by default to link in CI

### PR DESCRIPTION
**Pull request by driazati**
This adds an option to CMake to use `ld.gold` to link rather than `ld` (which symlinks to `ld.bfd` on Ubuntu by default). This shouldn't change any functionality, only a mild improvement on link times during builds (shaves off 1 minute) on CI.

Verify by searching for `ld.gold is available` in [the logs](https://circleci.com/api/v1.1/project/github/pytorch/pytorch/13046834/output/105/0?file=true&allocation-id=608c434338107e5b6cf938a1-0-build%2F7BDA2FF1)

Differential Revision: [D28123522](https://our.internmc.facebook.com/intern/diff/28123522/)

### CODE REVIEW

**Review by janeyx99 — APPROVED** [review]

**Comment by janeyx99** [comment]
Would it be worth also adding it to summary.cmake?

File: CMakeLists.txt [hunk 1]
```diff
@@ -684,6 +685,26 @@ if(NOT MSVC)
...
   endif()
+
+  # Use ld.gold if available, fall back to ld.bfd (the default ld) if not
+  if(USE_GOLD_LINKER)
+    if(USE_DISTRIBUTED AND USE_MPI)
+      # Same issue as here with default MPI on Ubuntu
+      # https://bugs.launchpad.net/ubuntu/+source/deal.ii/+bug/1841577
+      message(WARNING "Refusing to use gold when USE_MPI=1")
+    else()
+      execute_process(COMMAND ${CMAKE_C_COMPILER} -fuse-ld=gold -Wl,--version ERROR_QUIET OUTPUT_VARIABLE LD_VERSION)
```
Comment (walterddr) [inline CMakeLists.txt:696]: nit
```suggestion
      execute_process(
        COMMAND 
        "${CMAKE_C_COMPILER}" -fuse-ld=gold -Wl,--version 
         ERROR_QUIET 
         OUTPUT_VARIABLE LD_VERSION)
```

File: CMakeLists.txt [hunk 2]
```diff
@@ -684,6 +685,26 @@ if(NOT MSVC)
...
+      if(NOT "${LD_VERSION}" MATCHES "GNU gold")
+        message(WARNING "USE_GOLD_LINKER was set but ld.gold isn't available")
```
Comment (walterddr) [inline CMakeLists.txt:698]: we probably need to do either 
```
set(USE_GOLD_LINKER OFF)
```
or
```
caffe2_update_option(USE_GOLD_LINKER OFF)
```

**Review by walterddr — APPROVED** [review]
looks good overall. left a couple of comments. plz kindly take a look

**Comment by driazati** [comment]
MacOS failures look unrelated (`USE_GOLD_LINKER` is `OFF` in their logs, and linux test failure looks unrelated so going ahead with land

**Pull request closed without merging**